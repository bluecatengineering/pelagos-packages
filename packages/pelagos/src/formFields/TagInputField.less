@import '../../less/spacing';

.TagInputField {
	margin-top: @sp-12;
}
