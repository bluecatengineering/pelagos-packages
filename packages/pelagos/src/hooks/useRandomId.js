import {useMemo} from 'react';

/**
 * Returns either the specified string or a stable random string.
 * @param {string} [id] the ID to use.
 * @returns {string} the ID.
 */
const useRandomId = (id) =>
	useMemo(() => id || `e${performance.now().toString(36)}${Math.random().toString(36).slice(1)}`, [id]);

export default useRandomId;
