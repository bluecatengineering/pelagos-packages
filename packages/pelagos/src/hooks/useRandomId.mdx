{/* DO NOT EDIT, generated from useRandomId.js */}

import {Meta} from '@storybook/addon-docs/blocks';

<Meta title="hooks/useRandomId" />

# useRandomId

Returns either the specified string or a stable random string.

## Parameters

<table>
	<thead>
		<tr>
			<th>Param</th>
			<th>Type</th>
			<th>Description</th>
		</tr>
	</thead>
	<tbody>
		<tr>
			<td>id</td>
			<td>`string`</td>
			<td>the ID to use.</td>
		</tr>
	</tbody>
</table>

## Returns

`string` - the ID.
