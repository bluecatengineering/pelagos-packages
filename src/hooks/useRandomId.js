import {useMemo} from 'react';

export default (id) => useMemo(() => id || `e${('' + Math.random()).slice(2)}`, [id]);
