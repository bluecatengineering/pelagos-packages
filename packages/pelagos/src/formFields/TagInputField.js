import PropTypes from 'prop-types';

import TagInput from '../components/TagInput';
import useRandomId from '../hooks/useRandomId';

import FieldWrapper from './FieldWrapper';

/** A form field wrapper for [TagInput](/docs/components-taginput--docs). */
const TagInputField = ({id, className, label, required, tags, helperText, error, disabled, ...props}) => {
	id = useRandomId(id);
	const labelId = `${id}-label`;
	const helperId = `${id}-helper`;
	return (
		<FieldWrapper
			id={labelId}
			className={`TagInputField${className ? ` ${className}` : ''}`}
			htmlFor={id}
			label={label}
			required={required}
			helperId={helperId}
			helperText={helperText}
			error={error}>
			<TagInput
				{...props}
				id={id}
				tags={tags}
				error={error}
				disabled={disabled}
				aria-required={required}
				aria-labelledby={labelId}
				aria-describedby={helperId}
			/>
		</FieldWrapper>
	);
};

TagInputField.propTypes = {
	/** The component id. */
	id: PropTypes.string,
	/** The component class name(s). */
	className: PropTypes.string,
	/** The label text. */
	label: PropTypes.string.isRequired,
	/** Whether the field is required. */
	required: PropTypes.bool,
	/** The entered tags. */
	tags: PropTypes.array.isRequired,
	/** The default tags. */
	defaultTags: PropTypes.array,
	/** The tooltip for default tags. */
	defaultTooltipText: PropTypes.string,
	/** Additional information for the field. */
	helperText: PropTypes.string,
	/** The error text. */
	error: PropTypes.string,
	/** Whether the field is disabled. */
	disabled: PropTypes.bool,
	/** Function invoked to validate each tag. */
	validate: PropTypes.func.isRequired,
	/** Function invoked when the tags change. */
	onChange: PropTypes.func.isRequired,
	/** Function invoked when an error is detected. */
	onError: PropTypes.func.isRequired,
};

export default TagInputField;
