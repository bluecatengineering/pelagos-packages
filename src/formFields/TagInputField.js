import PropTypes from 'prop-types';

import TagInput from '../components/TagInput';
import LabelLine from '../components/LabelLine';
import useRandomId from '../hooks/useRandomId';

import FieldError from './FieldError';
import './TagInputField.less';

/** A tag input field. */
const TagInputField = ({id, className, label, optional, tags, error, ...props}) => {
	id = useRandomId(id);
	const labelId = `${id}-label`;
	return (
		<div className={`TagInputField${className ? ' ' + className : ''}`}>
			<LabelLine id={labelId} htmlFor={id} text={label} optional={optional && tags.length === 0} />
			<TagInput {...props} id={id} tags={tags} error={error} aria-labelledby={labelId} />
			<FieldError text={error} />
		</div>
	);
};

TagInputField.propTypes = {
	/** The component id. */
	id: PropTypes.string,
	/** The component class name(s). */
	className: PropTypes.string,
	/** The label text. */
	label: PropTypes.string.isRequired,
	/** Whether the field is optional. */
	optional: PropTypes.bool,
	/** The entered tags. */
	tags: PropTypes.array.isRequired,
	/** The default tags. */
	defaultTags: PropTypes.array,
	/** The tooltip for default tags. */
	defaultTooltipText: PropTypes.string,
	/** The error text. */
	error: PropTypes.string,
	/** Function invoked to validate each tag. */
	validate: PropTypes.func.isRequired,
	/** Function invoked when the tags change. */
	onChange: PropTypes.func.isRequired,
	/** Function invoked when an error is detected. */
	onError: PropTypes.func.isRequired,
};

export default TagInputField;
