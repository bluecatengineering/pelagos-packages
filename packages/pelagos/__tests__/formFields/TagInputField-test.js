import {shallow} from 'enzyme';

import TagInputField from '../../src/formFields/TagInputField';
import useRandomId from '../../src/hooks/useRandomId';

jest.unmock('../../src/formFields/TagInputField');

useRandomId.mockReturnValue('random-id');

describe('TagInputField', () => {
	describe('rendering', () => {
		it('renders expected elements', () => {
			const wrapper = shallow(
				<TagInputField
					id="test"
					label="Label"
					required
					tags={['foo']}
					helperText="Helper text"
					validate={jest.fn()}
					onChange={jest.fn()}
					onError={jest.fn()}
				/>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
			expect(useRandomId.mock.calls).toEqual([['test']]);
		});

		it('renders expected elements when className is set', () => {
			const wrapper = shallow(
				<TagInputField
					id="test"
					className="TestClass"
					label="Label"
					tags={['foo']}
					validate={jest.fn()}
					onChange={jest.fn()}
					onError={jest.fn()}
				/>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders expected elements if error is set', () => {
			const wrapper = shallow(
				<TagInputField
					id="test"
					label="Label"
					tags={['foo']}
					helperText="Helper text"
					error="Error"
					validate={jest.fn()}
					onChange={jest.fn()}
					onError={jest.fn()}
				/>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});
	});
});
