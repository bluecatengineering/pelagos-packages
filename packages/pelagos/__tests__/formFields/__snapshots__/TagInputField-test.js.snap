// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`TagInputField rendering renders expected elements 1`] = `
<FieldWrapper
  className="TagInputField"
  helperId="random-id-helper"
  helperText="Helper text"
  htmlFor="random-id"
  id="random-id-label"
  label="Label"
  required={true}
>
  <TagInput
    aria-describedby="random-id-helper"
    aria-labelledby="random-id-label"
    aria-required={true}
    id="random-id"
    onChange={[MockFunction]}
    onError={[MockFunction]}
    tags={
      [
        "foo",
      ]
    }
    validate={[MockFunction]}
  />
</FieldWrapper>
`;

exports[`TagInputField rendering renders expected elements if error is set 1`] = `
<FieldWrapper
  className="TagInputField"
  error="Error"
  helperId="random-id-helper"
  helperText="Helper text"
  htmlFor="random-id"
  id="random-id-label"
  label="Label"
>
  <TagInput
    aria-describedby="random-id-helper"
    aria-labelledby="random-id-label"
    error="Error"
    id="random-id"
    onChange={[MockFunction]}
    onError={[MockFunction]}
    tags={
      [
        "foo",
      ]
    }
    validate={[MockFunction]}
  />
</FieldWrapper>
`;

exports[`TagInputField rendering renders expected elements when className is set 1`] = `
<FieldWrapper
  className="TagInputField TestClass"
  helperId="random-id-helper"
  htmlFor="random-id"
  id="random-id-label"
  label="Label"
>
  <TagInput
    aria-describedby="random-id-helper"
    aria-labelledby="random-id-label"
    id="random-id"
    onChange={[MockFunction]}
    onError={[MockFunction]}
    tags={
      [
        "foo",
      ]
    }
    validate={[MockFunction]}
  />
</FieldWrapper>
`;

exports[`TagInputField rendering renders expected elements when disabled 1`] = `
<FieldWrapper
  className="TagInputField"
  helperId="random-id-helper"
  helperText="Helper text"
  htmlFor="random-id"
  id="random-id-label"
  label="Label"
  required={true}
>
  <TagInput
    aria-describedby="random-id-helper"
    aria-labelledby="random-id-label"
    aria-required={true}
    disabled={true}
    id="random-id"
    onChange={[MockFunction]}
    onError={[MockFunction]}
    tags={
      [
        "foo",
      ]
    }
    validate={[MockFunction]}
  />
</FieldWrapper>
`;
